Modules/POSIX/Sources/POSIX/Regex.swift

```diff
@@ -0,0 +1,308 @@
+        let result = regexec(&preg, string, regexMatches.count, &regexMatches, options.rawValue)
+
+        if result != 0 {
+            return false
+        }
+
+        return true
+    }
+
+    public func groups(in string: String, options: MatchOptions = []) -> [String] {
+        var string = string
+        let maxMatches = 10
+        var groups = [String]()
+        var regexMatches = [regmatch_t](repeating: regmatch_t(), count: maxMatches)
+
+        // Iterate over the string per batch of 10 matches
+        while true {
+            let result = regexec(&preg, string, regexMatches.count, &regexMatches, options.rawValue)
+
+            if result != 0 {
+                break // Unmatched regex
+            }
+
+            if regexMatches[0].rm_eo == regexMatches[0].rm_so {
+                break // matches the empty string: infinite loop
+            }
+
+            var groupIdx = 1
+
+            // Iterate over the matches
+            while regexMatches[groupIdx].rm_so != -1 {
+                let group = (start: regexMatches[groupIdx].rm_so, end: regexMatches[groupIdx].rm_eo)
+
+                // Use UTF8View for unicode regexes
+                let startIndexUTF8 = index(from: group.start, in: string.utf8)
+                let endIndexUTF8 = index(from: group.end, in: string.utf8)
+
+                let match = String(string.utf8[startIndexUTF8..<endIndexUTF8])!
+                groups.append(match)
+                groupIdx += 1
+            }
+
+            let indexOfEndOfMatchUTF8 = index(from: regexMatches[0].rm_eo, in: string.utf8)
+            let remainderString = String(string.utf8[indexOfEndOfMatchUTF8..<string.utf8.endIndex])!
+
+            if !remainderString.isEmpty {
+                string = remainderString
+            } else {
+                break
+            }
+        }
+
+        return groups
+    }
+
+    public func replace(in string: String, with template: String, options: MatchOptions = []) -> String {
+        var string = string
+        let maxMatches = 10
+        var totalReplacedString: String = ""
+        let templateArray = [UInt8](template.utf8)
+
+        while true {
+            var regexMatches = [regmatch_t](repeating: regmatch_t(), count: maxMatches)
+            let result = regexec(&preg, string, regexMatches.count, &regexMatches, options.rawValue)
+
+            if result != 0 {
+                break // Unmatched regex
+            }
+
+            if regexMatches[0].rm_eo == regexMatches[0].rm_so {
+                break // matches the empty string: infinite loop
+            }
+
+            let start = Int(regexMatches[0].rm_so)
+            let end   = Int(regexMatches[0].rm_eo)
+            var replacedStringArray = [UInt8](string.utf8)
+            replacedStringArray.replaceSubrange(start..<end, with: templateArray)
+
+            guard let replacedString = String(data: Data(replacedStringArray), encoding: .utf8) else {
```

Comment:
This is the reason why `Foundation` is used. Should I try to find an alternative?